protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,489 @@
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        voteToken = IStakedTempleVoteToken(_voteToken);
+    }
+
+    /**
+     * @notice Set starter of rewards distribution for the next epoch
+     * @dev If starter is address zero, anyone can call `distributeRewards` to apply and 
+     *      distribute
+      rewards for next 7 days
+     * @param _starter Starter address
+     */
+    function setDistributionStarter(address _starter) external onlyElevatedAccess {
+        /// @notice Starter can be address zer0
+        distributionStarter = _starter;
+        emit DistributionStarterSet(_starter);
+    }
+
+    /**
+     * @notice Set migrator
+     * @param _migrator Migrator
+     */
+    function setMigrator(address _migrator) external override onlyElevatedAccess {
+        if (_migrator == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        migrator = _migrator;
+        emit MigratorSet(_migrator);
+    }
+
+    /**
+     * @notice Set reward distribution cooldown
+     * @param _cooldown Cooldown in seconds
+     */
+    function setRewardDistributionCoolDown(uint160 _cooldown) external override onlyElevatedAccess {
+        /// @dev zero cooldown is allowed
+        rewardDistributionCoolDown = _cooldown;
+        emit RewardDistributionCoolDownSet(_cooldown);
+    }
+
+    /**
+     * @notice Set half time parameter for calculating vote weight.
+     * @dev The voting half-time variable determines the time it takes until half the voting weight is reached for a stake.
+     *      Formular from st-yETH https://docs.yearn.fi/getting-started/products/yeth/overview
+     * @param _halfTime Cooldown in seconds
+     */
+    function setHalfTime(uint256 _halfTime) external override onlyElevatedAccess {
+        if (_halfTime == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        halfTime = _halfTime;
+        emit HalfTimeSet(_halfTime);
+    }
+
+    /**
+      * @notice For migrations to a new staking contract if TGLD changes
+      *         1. Withdraw `staker`s tokens to the new staking contract (the migrator)
+      *         2. Any existing rewards are claimed and sent directly to the `staker`
+      * @dev Called only from the new staking contract (the migrator).
+      *      `setMigrator(new_staking_contract)` needs to be called first
+      * @param staker The staker who is being migrated to a new staking contract.
+      * @param amount The amount to migrate - generally this would be the staker's balance
+      */
+    function migrateWithdraw(address staker, uint256 amount) external override onlyMigrator {
+        _withdrawFor(staker, msg.sender, amount, true, staker);
+    }
+    
+    /**
+     * @notice Distributed TGLD rewards minted to this contract to stakers
+     * @dev This starts another 7-day rewards distribution. Calculates new `rewardRate` from any left over rewards up until now
+     */
+    function distributeRewards() external {
```

Comment:
I think this should also have the `updateReward()` modifier?

I'd think we want to ensure the `rewardPerTokenStored` is updated to the latest prior to updating the rates and `lastUpdateTime`

Reply:
good catch